### ISSUE TITLE
QueryTraverser only descends into FragmentSpreads once

### ISSUE THREAD

**Issue by kaqqao** [graphql-java/graphql-java#1695]
When the same fragment is used multiple times, `QueryTraverser` descends into its field only once.

A simple example:

```graphql
{
  users {
    ...userInfo
    friends {
      ...userInfo
    }
  }

  fragment userInfo on User {
    name
  }
}
```

On a query such as this, `QueryTraverser` will visit the `name` field only once (and will visit the `...userInfo` spread twice, as expected). This is distinctly different from the behavior one would get for a semantically equivalent query with no fragments:

```graphql
{
  users {
    name
    friends {
      name
    }
  }
}
```
One immediate implication is on query complexity analysis. A (malicious?) query with recursive fragment spreads will get a score underestimated by orders of magnitude.

As an example, `MaxQueryComplexityInstrumentation` will calculate the complexity of the full introspection query to be 45, while in reality over a 100 fields are requested. And that query isn't intentionally crafted to trick the calculation with large, recursive, fragment spreads.

### Investigation

The reason why this happens is that `QueryTraverser` [in this line](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/QueryTraverser.java#L158) considers `FragmentDefinition` the only child of each `FragmentSpread`. Since both spreads have the same definitions, when the traverser encounters the 2nd spread, it will falsely believe ([in this line](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/util/DefaultTraverserContext.java#L149)) its children (the definition) have already been visited and will not descend.

Happy to help fixing this, but I'd need guidance on the approach...

**Comment by andimarek** [comment]
@kaqqao thanks for reporting.

That is indeed an interesting case: The general traverser doesn't visit any vertex in a graph twice in order to avoid loops and extra work.

But I can see that in the case it would be good to visit the FragmentDefinition twice (or as many times it is referenced).

**Comment by andimarek** [comment]
@kaqqao to add on this: what is your expectations around duplicate fields?

for example:

```graphql
{ foo { bar} foo {bar}} 
```
would you expect that `bar` is visited twice or once?

**Comment by kaqqao** [comment]
I understood the traverser ignores duplicates and I found that reasonable enough, especially for the purposes of complexity analysis (as the "price" of resolving each field will only be paid once). But fragments I think are special, as they can be spread in unrelated contexts, effectively expanding into entirely new (and not visited) fields. Of course, spreading a fragment can result in duplicate fields, going back to your original question... which I'm still unsure how to answer.
On one hand, it makes sense for the traverser to visit each field as specified. But since the duplicate fields get merged during resolution anyway, visiting them only once sounds correct too. I'd honestly go for whatever is easier to implement and see if anyone has a legit use-case for the alternative.

If I said something nonsensical, please forgive, it's 3am where I am and my brain isn't at it's best 😶 I'll think some more on this in the morning.

**Comment by andimarek** [comment]
@kaqqao I created a new PR which adds an analyzer which produces a "query execution tree": this tree contains all the fields merged together.

Have a look at this test: 
https://github.com/graphql-java/graphql-java/pull/1731/files#diff-72f87f8a5876e2927c1df5eb8cb4bd38R479

A complexity analysis should probably traverse this query execution tree for the most accurate analysis.

**Comment by kaqqao** [comment]
`NormalizedQueryTree` looked very promising for complexity analysis, but can't seem to provide the info I'm looking for after all. The main hurdle is that I'd like to detect alternative subtrees and pick the most expensive one.

E.g. for a query like this:

```graphql
{
    pet(id: 123) {
            sound
            owner {
                name
            }
            ... on Cat {
                catField
            }
            ... on Dog {
                dogField
                owner {
                    age
                }
            }
      }
}
```

I'd like to detect that `... on Cat` and `...on Dog` subtrees are alternatives, but that doesn't seem possible by looking at `NormalizedField`s.

I'm currently using a a tangled mess of internal methods to calculate complexity, and on every upgrade I'm terrified of anything changing in there. Do you see anything I could use instead for this purpose?

**Comment by bbakerman** [comment]
> I'd like to detect that ... on Cat and ...on Dog subtrees are alternatives, calculate the complexity for both and pick only the most expensive one, but that doesn't seem possible by looking at NormalizedField structure.

I think we need to understand more about that.  `NormalisedField` allows you to know the object type (so you can know if its a Cat or a Dog) and you can see the parent and child fields from a given `NF`

What is it that is missing for you?

**Comment by jstlnd** [comment]
Hello!
Is there a solution coming for this issue? We realized we're accepting queries that have way higher complexity than we should because of fragments being used in several places.
Cheers

**Comment by kaqqao** [comment]
Has anything changed since 2020 regarding this? The `NormalizedQueryTree` PR was never merged, so that avenue is closed... Is there an alternative these days?

**Comment by bbakerman** [comment]
The `ExecutableNormalizedXXX` classes will be made public in the next version.  We started off internal as we built them out

They actually power quite a bit of the Atlassian graphql code.  Nadel (our federation library) is based entirely on ENFs and we recently used it in some schema CLI work as well as some subscription work.

So in short they have proven themselves.  So we will at some point soon make them public API.

So this is the way forward.  I like your PR linked as an idea.  See the PR for more comments.
